### DISCUSSION TITLE
Wishing list: Pulsar SQL should support readCompact on topics

### DISCUSSION THREAD

**Discussion by Hugues514** [apache/pulsar#18764] [General]
So far, when a Pulsar topic is compacted, the Pulsar SQL query will read all the records, not considering the compacted option

It could be interesting to be able to query such a topic using a readCompact option so that the reader will skip all the 'compacted records'.

**Comment by jiazhai** [comment]
We need a special handle in `PulsarSplitManager.getSplitsForTopic` to split based also on `compactionHorizon`(contained in `CompactedTopicImpl`); so that for data before compactionHorizon, we could read from compacted topic, while data after compactionHorizon, read from non-compacted topic.

**Comment by jiazhai** [comment]
more info: 
https://github.com/apache/pulsar/wiki/PIP-19:-Pulsar-SQL
https://github.com/apache/pulsar/wiki/PIP-14:-Topic-compaction

**Comment by pointearth** [comment]
Can you guys open this ticket and finish it or move it to a suitable appointment? appreciate!
[Add readCompact option #7461](https://github.com/apache/pulsar/pull/7461 )

**Comment by longtengz** [comment]
Does this also imply pulsar sql cannot process empty payload messages for the time being? Because I'm experiencing #13127

**Comment by tisonkun** [comment]
Moved to general discussions since there's a wishing list item and no one works on it now.
